azurerm/internal/services/resource/resource_arm_provider_registration.go

```diff
@@ -0,0 +1,188 @@
+package resource
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"time"
+
+	"github.com/Azure/azure-sdk-for-go/profiles/2017-03-09/resources/mgmt/resources"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/resource"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/helpers/azure"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/clients"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/timeouts"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/utils"
+)
+
+func resourceArmResourceProviderRegistration() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceArmResourceProviderCreateUpdate,
+		Read:   resourceArmResourceProviderRead,
+		Update: resourceArmResourceProviderCreateUpdate,
+		Delete: resourceArmResourceProviderDelete,
+		Importer: &schema.ResourceImporter{
+			State: schema.ImportStatePassthrough,
+		},
+
+		Timeouts: &schema.ResourceTimeout{
+			Create: schema.DefaultTimeout(30 * time.Minute),
+			Read:   schema.DefaultTimeout(5 * time.Minute),
+			Update: schema.DefaultTimeout(30 * time.Minute),
+			Delete: schema.DefaultTimeout(30 * time.Minute),
+		},
+
+		Schema: map[string]*schema.Schema{
+			"resource_provider_namespace": {
```

Comment:
I would say the user should implement the for_each on the resource in case of multiple RPs.

variable "provider_namespaces" {
  type    = list(string)
  default = ["Microsoft.Management"]
}

resource "azurerm_resource_provider_registration" "example" {
  for_each = toset(var.provider_namespaces)

  resource_provider_namespace = each.value
}

The underlying API does not support taking the list of RPs as well.

Reply:
@Humoiz, I think what @katbyte is asking for here is something like this:

```
"resource_providers": {
	Type:     schema.TypeList,
	Required: true,
	MaxItems: 50,
	Elem: &schema.Schema{
		Type: schema.TypeString,
		ValidateFunc: validation.<YourCustomValidationFunctionHere>,
	},
},
```
That way when you define the resource you can pass the resource providers like this:
```
resource "azurerm_resource_provider_registration" "example" {
  resource_providers = ["Microsoft.Management", "Microsoft.Network", "Microsoft.Storage"]
}
```
In your create function you will have to loop over the names in the slice and call the API individually for each of the resource providers that the user wants to have registered. Personally, I would do a get on the provider and check to see if it's registered or not, if it is not call the API to register it, if there is a failure add the error to a error slice and then move to the next provider that needs to be registered. Once you have walked the entire list of resource providers check to see if the error slice is empty or not, if not return the error slice else all registration calls were successful call the read function and exist with a success. Does that make sense?

Reply:
I'd disagree here - as @Humoiz has mentioned we can use a foreach on the resource to manage individual resources

From Terraform's perspective it's better to have a resource per RP - particularly when longer term there's features which can be enabled/disabled here - which could make sense within this resource; so I think this is fine as-is 👍